drivers: move gpio_keys from gpio to input

fabiobaltieri opened the pull request:
First input driver ported to the input subsystem. This used to live in the gpio directory but was using a custom APIs so I think it should be fine to just convert it over. Ported driver and tests in a single commit to keep the thing bisectable.

Also adding an overlay to test it in the `input_dump` sample, I'd imagine long term we could change the board dts themselves but for now this is enough to show how to use the driver.

---

Port the gpio_keys_zephyr driver from the gpio subsystem with a dedicated API to the input subsystem reporting input events.

Move the test as well, simplify the cases a bit since the API is simpler now.

fabiobaltieri committed:
drivers: move gpio_keys from gpio to input

Port the gpio_keys_zephyr driver from the gpio subsystem with a
dedicated API to the input subsystem reporting input events.

Move the test as well, simplify the cases a bit since the API is simpler
now.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
samples: input_dump: add a board overlay for testing zephyr,gpio-keys

Add an initial board overlay in the input_dump directory for testing
zephyr,gpio-keys. This takes the exsiting "gpio-keys" node, changes the
compatible to the one used by the gpio keys input driver and set some
codes.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

nordicjm commented:
What is the difference in output file size for this if you compile with the old GPIO way vs the new input way on e.g. `frdm_k64f`?

fabiobaltieri commented:
> What is the difference in output file size for this if you compile with the old GPIO way vs the new input way on e.g. `frdm_k64f`?

You mean with the custom API (before this PR) and after? Let's see, trying on nrf52dk since that's what I'm using, for the "before" case I hacked this on top of the blinky sample: https://gist.github.com/fabiobaltieri/2101a7e9eadb42f76251f54939424e94 and I get this:

```
$ west build -p -b nrf52dk_nrf52832 samples/basic/blinky
...
Memory region         Used Size  Region Size  %age Used
           FLASH:       21520 B       512 KB      4.10%
             RAM:        7104 B        64 KB     10.84%
```

with the input stuff in syncronous mode I get

```
$ west build -p -b nrf52dk_nrf52832 samples/subsys/input/input_dump -DCONFIG_INPUT_MODE_SYNCHRONOUS=y
Memory region         Used Size  Region Size  %age Used
           FLASH:       21456 B       512 KB      4.09%
             RAM:        7104 B        64 KB     10.84%
```

with the thread:

```
west build -p -b nrf52dk_nrf52832 samples/subsys/input/input_dump -DCONFIG_INPUT_MODE_THREAD=y
...
Memory region         Used Size  Region Size  %age Used
           FLASH:       22184 B       512 KB      4.23%
             RAM:        8064 B        64 KB     12.30%
```

Guess the few saved bytes are from the removed interrupt disable code.

fabiobaltieri committed:
dts: bindings: zephyr,gpio-keys require code and add docs

The zephyr,gpio-keys is now emitting input events, so it makes sense to
require a key code to be set. Change the zephyr,code property to be
required and add an example in the binding description.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

henrikbrixandersen reviewed (changes requested):
~Please leave the existing GPIO keys as-is. They provide a no-overhead, simple solution used by _a lot_ of applications.~

~I'm fine with adding a similar driver for passing GPIO key interrupts as input events, but let's not force users to have the input subsystem enabled to solve something that is already solved by the GPIO driver itself.~

~This is similar to having the option of using an LED driver for a simple GPIO connected LED, but not forcing users to use a dedicated driver class on top of the GPIO driver if not needed.~

Nevermind me. I didn't even realise we already _have_ a dedicated GPIO keys driver, sigh.

cfriedt reviewed (commented):
Just minor changes to copyrights please

fabiobaltieri commented on tests/drivers/input/gpio_keys/boards/native_posix.overlay:
```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright (c) 2020 Friedt Professional Engineering Services, Inc
```
Oh I just copied the file, don't even know why we bother to be honest, it'd be so much less of a hassle if we just had some "The Zephyr Project Authors" like we do with ChromiumOS. I'll change it.

fabiobaltieri commented:
wait - somehow I got a bunch of unrelated stuff in the last push

fabiobaltieri commented:
@henrikbrixandersen @cfriedt can you take another look? Thanks!
